proposals/lip-0055.md

```diff
@@ -105,16 +115,42 @@ The full processing of a block is organized as follows.
-12. **After command execution**: Each module can define protocol logic that is processed *after* the command has been executed.
+8. **Transaction verification**: Each module can define protocol logic that verifies a transaction, possibly by accessing the state store. If an error occurs, the transaction is invalid and the whole block is discarded.
+9. **Command verification**: The command corresponding to the `moduleID`-`commandID` combination is verified. If an error occurs, the transaction is invalid and it is not included in the block.
+10. **Before command execution**: Each module can define protocol logic that is processed *before* the command has been executed. If an error occurs, the transaction is invalid and the whole block is discarded.
+11. **Command execution**: The command corresponding to the `moduleID`-`commandID` combination is executed. If an error occurs, the transaction is failed and the processing continues.
+12. **After command execution**: Each module can define protocol logic that is processed *after* the command has been executed. If an error occurs, the transaction is invalid and the whole block is discarded.
 13. **After transactions execution**: Each module can define protocol logic that is executed *after* all the transactions contained in the block have been processed.
 14. **Result verification**: Block header properties, which require accessing the state store *after* all state transitions implied by the block have been executed, are verified.
 15. **Block storage**: The block is persisted into the database.
 16. **Peers notification**: Other peers in the P2P network are notified of the new block.
 
+
+### Block Generation Stages
+
+<img src="lip-0055/generation_stages.png" width="70%">
+
+*Figure 2: A schematic depiction of the various stages of a block generation. The steps performed in the generator domain are indicated by blue boxes, while the steps performed in the state-machine domain are indicated by red boxes. Within the state-machine domain, the transactions processing is indicated by purple boxes. The network domain is indicated by green boxes.*
+
+In this section, we describe the various stages of a block generation.
+
+Similarly to the block processing, the block generation is split between the network, consensus, and state-machine domains.
+The full generation of a block is organized as follows.
+
+1. **Header creation**: Block header properties that require access to the state store *before* any state transitions implied by the block are executed are inserted in this stage.
+2. **Block initialization**: Block assets properties that require access to the state store *before* any state transitions implied by the block are executed are inserted in this stage.
+3. **Before transactions execution**: Each module can define protocol logic that is executed *before* the transactions contained in the block are processed.
```

Comment:
Above, we have hooks/steps for every piece of data added to a block (header, asset), but we somehow omit when/how transactions are added to the block (before they are executed here). I think it would be good to mention this somewhere.

Reply:
Mh, I'm not sure how it works in practice. Are transactions preselected or are they added one by one after the "before transactions execution" stage is completed? @shuse2 can you help me?

Reply:
The process is like
1. verify transaction
2. take state snapshot
3. before command execute, execute command, after command execute
4. add to block or restore snapshot and not include the transaction

Reply:
Added a paragraph in stage 3

Reply:
Can be resolved.